# ２量子ビットの計算
量子コンピュータでは、複数の量子ビットの操作により、複雑な計算を行うことができます。
まずは、Blueqatをインストールし、ツールを読み込みましょう。

In [1]:
!pip install blueqat
from blueqat import Circuit

numba 0.49.0 has requirement numpy>=1.15, but you'll have numpy 1.14.6 which is incompatible.
You are using pip version 10.0.1, however version 20.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


## 量子回路の作成
量子回路の作成方法は１量子ビットの場合と同じです。CXゲートを使ってみます。小文字でcxと書き、カッコの中に、0と1をカンマで繋げて記入します。そして、測定を行うmをかいて、runで100回実行します。

In [2]:
Circuit().cx[0,1].m[:].run(shots=100)

Counter({'00': 100})

00という答えが100回出ました。CXゲートは、「コントロールビット」と「ターゲットビット」を順番に指定します。これは、コントロールビットの値が1の時だけ、ターゲットビットにXゲートをかけます。

今回は、量子ビットは、両方とも0からスタートして、cxゲートがかかりました。コントロールビットが0なので、ターゲットビットには何もおきません。回路は以下のようになります。

```
0 ---*---
     |
0 ---X---
```

次にコントロールビットにXゲートをかけて、0を1にしてから、CXゲートをかけてみます。

In [4]:
Circuit().x[0].cx[0,1].m[:].run(shots=1)

Counter({'11': 1})

今度は、11という答えが出ました。コントロールビットが1のため、ターゲットビットにもXゲートがかかり、0から1になりました。このように2つの量子ビットを指定して計算を行います。

## （応用）量子ビットの数を指定する
量子ビットの数を最初から指定することもできます。その際にはCircuit()のなかに引数を設定しましょう。

In [5]:
Circuit(3).x[0].cx[0,1].m[:].run(shots=1)

Counter({'110': 1})

110となりました。最初の2量子ビットは計算に使用されましたが、最後の量子ビットは使用されないままでした。以上です。

# 概要
ここでは２量子ビットの計算を説明します。

## 量子ビット
２量子ビットの場合は 00, 01, 10, 11 は以下のような状態ベクトルで表します。

<img src="./img/002/002_02_0.png" width="50%">

00 の状態ベクトルはテンソル積を用いると以下のように計算できます。

<img src="./img/002/002_02_1.png" width="50%">

## CX回路
CXゲートはコントロールビットが1ならばターゲットビットにXゲートを施すものです。

CXゲートは以下のような行列です。

<img src="./img/002/002_02_2.png" width="20%">

00 にCXゲートを施すと以下のようになります。

<img src="./img/002/002_02_3.png" width="43%">

この場合コントロールビットは0なので何も変化しません。

次に 10 にCXゲートを施すと以下のようになります。

<img src="./img/002/002_02_4.png" width="43%">

この場合コントロールビットは1なのでターゲットビットが反転して1になっているのがわかります。

## 観測
２量子ビットの場合の観測は 00, 01, 10, 11 が確率的に観測されます。   
それぞれ出力される確率は１量子ビットの場合と同じように各係数の絶対値の2乗となります。

例えば以下のような状態ベクトルを考えます。

<img src="./img/002/002_02_5.png" width="40%">

これを1回観測すると 00, 01, 10, 11 が出力される確率は 1/4 となります。

## 2量子ビット以上の場合
３量子ビットの場合で最初の２量子ビットにCXゲートをかけるとします。
100 にCXを施したとき計算は以下のようになります。

<img src="./img/002/002_02_6.png" width="45%">

上の式から最後のビットには何も変化しない行列Iを施して計算していることがわかります。